### Assembling the model

In [47]:
import numpy as np 
import tensorflow as tf
import pandas as pd

npz = np.load('titanic_train_data.npz', allow_pickle=True)
train_inputs, train_targets = npz['inputs'], npz['targets']

npz = np.load('titanic_validation_data.npz', allow_pickle=True)
validation_inputs, validation_targets = npz['inputs'], npz['targets']

npz = np.load('titanic_test_data.npz', allow_pickle= True)
test_data = npz['inputs']
npz = np.load('titanic_test_PassengerId.npz', allow_pickle= True)
test_PassengerId = npz['inputs']

### Model
Outline, Optimizers, Loss, Early Stopping and Training

In [48]:
input_size = train_inputs.shape[1]
hidden_layer_size = 100
output_size = 1

model = tf.keras.Sequential([
                                tf.keras.layers.Dense(hidden_layer_size,activation='tanh'),
                                tf.keras.layers.Dense(hidden_layer_size,activation='tanh'),
                                tf.keras.layers.Dense(hidden_layer_size,activation='tanh'),
                                tf.keras.layers.Dense(output_size,activation='sigmoid'),
                            ])

model.compile(optimizer='adam',loss='binary_crossentropy', metrics = ['accuracy'])

batch_size = 100
max_epochs = 100

early_stopping = tf.keras.callbacks.EarlyStopping(patience= 2)

model.fit(
            train_inputs,
            train_targets,
            batch_size= batch_size,
            epochs = max_epochs,
            callbacks= [early_stopping],
            validation_data= (validation_inputs,validation_targets),
            verbose = 2
         )

Epoch 1/100
7/7 - 1s - loss: 0.5861 - accuracy: 0.7480 - val_loss: 0.5092 - val_accuracy: 0.7826 - 821ms/epoch - 117ms/step
Epoch 2/100
7/7 - 0s - loss: 0.4758 - accuracy: 0.7967 - val_loss: 0.5076 - val_accuracy: 0.7826 - 52ms/epoch - 7ms/step
Epoch 3/100
7/7 - 0s - loss: 0.4632 - accuracy: 0.8065 - val_loss: 0.5169 - val_accuracy: 0.7826 - 48ms/epoch - 7ms/step
Epoch 4/100
7/7 - 0s - loss: 0.4542 - accuracy: 0.8081 - val_loss: 0.5107 - val_accuracy: 0.7971 - 55ms/epoch - 8ms/step


We get 79.7% accuracy, which is good. There's so much uncertainty, due to that "noise", but also due to other factors that are not given in the table, like people's health conditions, personality traits, personal relations and so on, that it's completely impossible to predict survival from the given data alone.

In [50]:
survived = model.predict(test_data)
survived_binary = (survived >= 0.5).astype(int)

Answer['']= 

14/14 [==============================] - 0s 601us/step


0
0  PassengerId